copy/copy.go

```diff
@@ -851,6 +979,49 @@ func (c *copier) copyBlobFromStream(ctx context.Context, srcStream io.Reader, sr
 		compressionOperation = types.PreserveOriginal
 		inputInfo = srcInfo
 	}
+	if decrypted {
+		inputInfo.MediaType = srcInfo.MediaType
+	}
+
+	// Perform image encryption for valid mediatypes if encryptConfig provided
+	var (
+		encryptMediaType string
+		encrypted        bool
+		finalizer        ocicrypt.EncryptLayerFinalizer
+	)
+	if toEncrypt {
+		switch srcInfo.MediaType {
+		case manifest.DockerV2Schema2LayerMediaType, ocispec.MediaTypeImageLayerGzip:
+			encryptMediaType = manifest.DockerV2Schema2LayerGzipEncMediaType
+		case ocispec.MediaTypeImageLayer:
+			encryptMediaType = manifest.DockerV2Schema2LayerEncMediaType
+		}
+
+		if encryptMediaType != "" && c.encryptConfig != nil {
+			var annotations map[string]string
+			if !decrypted {
+				annotations = srcInfo.Annotations
+			}
+			desc := ocispec.Descriptor{
+				MediaType:   srcInfo.MediaType,
+				Digest:      srcInfo.Digest,
+				Size:        srcInfo.Size,
+				Annotations: annotations,
+			}
+
+			s, fin, err := ocicrypt.EncryptLayer(c.encryptConfig, destStream, desc)
+			if err != nil {
+				return types.BlobInfo{}, errors.Wrapf(err, "Image encryption failed for the digest %+v", srcInfo.Digest)
+			}
+
+			destStream = s
+			finalizer = fin
+			inputInfo.Digest = ""
+			inputInfo.Size = -1
+			inputInfo.MediaType = encryptMediaType
```

Comment:
Ah, so we're setting a `DockerV2Schema2` prefix to an OCI mime type. I very much prefer to use the OCI prefix for it.